uptime_service_validation/database/create_tables.sql

```diff
@@ -68,9 +69,54 @@ CREATE TABLE IF NOT EXISTS score_history (
 	node_id INT,
 	score_at TIMESTAMP(6), 
 	score INT, 
-	score_percent NUMERIC(10,2),
+	score_percent NUMERIC(5,2),
 	CONSTRAINT fk_nodes
 		FOREIGN KEY(node_id) 
 		REFERENCES nodes(id)
 );
-CREATE UNIQUE INDEX IF NOT EXISTS uq_sh_node_score_at ON score_history USING btree (node_id, score_at);
\ No newline at end of file
+CREATE UNIQUE INDEX IF NOT EXISTS uq_sh_node_score_at ON score_history USING btree (node_id, score_at);
+
+-- Table creation for points_summary
+-- The points_summary table aggregates data related to node scoring.
+-- It is designed to hold a unique combination of bot_log_id and node_id.
+-- This allows us to ensure that each node's contribution within a single batch
+-- is counted once, preventing the final score from exceeding 100% in cases
+-- where block producers may send more than one valid submission in a batch.
+-- It is used by `update_scoreboard` function to update the final_score of each node.
+CREATE TABLE IF NOT EXISTS points_summary (
+    bot_log_id INT NOT NULL,
+    node_id INT NOT NULL
+);
+
+-- Unique index for points_summary
+-- A unique index on bot_log_id and node_id ensures that the pair is unique within the table.
+-- This supports the trigger function's role in preventing duplicate summary entries
+-- for the same bot_log and node combination.
+CREATE UNIQUE INDEX IF NOT EXISTS uq_ps_bot_log_node ON points_summary USING btree (bot_log_id, node_id);
+
+-- Trigger function definition for updating the points_summary
+-- This function, fn_update_point_summary, is a trigger function that responds to insert operations.
+-- When a new point is inserted, this function attempts to insert a corresponding entry into
+-- the points_summary table. If an entry with the same bot_log_id and node_id already exists,
+-- it does nothing (avoids duplicate entries), ensuring that each node's score is calculated
+-- correctly for a batch and that the final_score does not exceed 100%.
+CREATE OR REPLACE FUNCTION fn_update_point_summary()
+RETURNS trigger
+LANGUAGE plpgsql
+AS $function$
+BEGIN
+    INSERT INTO points_summary(bot_log_id, node_id)
+    VALUES(NEW.bot_log_id, NEW.node_id)
+    ON CONFLICT (bot_log_id, node_id) DO NOTHING;
+    RETURN new;
+END;
+$function$;
```

Comment:
Wouldn't it be better done in Python code? As our story shows, this method of doing things is obscure, untestable and easy to miss when making changes or investigating the results. If it's easiest to do as a Postgres trigger, let's at least mention it in READMEs so that future generations are not surprised by its existence like we were. Let's also mention it in comments near the Python code which triggers this function.

Reply:
From talking to Umesh I understood there were performance consideration so they decided to go with the DB route, I tried to document throughly in the comments, but will add a note in the README too.